### PR TITLE
[new release] multicont (1.0.1)

### PR DESCRIPTION
**Pull request by dhil**
Multi-shot continuations in OCaml

- Project page: <a href="https://github.com/dhil/ocaml-multicont">https://github.com/dhil/ocaml-multicont</a>

##### CHANGES:

This release is a purely administrative release which ports the build infrastructure to [dune](https://github.com/ocaml/dune) in order to resolve the reported build issues (e.g. https://github.com/ocaml/opam-repository/pull/23972).

### CODE REVIEW

**Comment by avsm** [comment]
Thanks for taking the trouble to fix the build @dhil! 

Should we be doing something in OCaml 5.1 or 5.2 to expose `multicont_fiber_id` in the C runtime?  Exposing it there seems reasonable (as opposed to from the ML end), but I don't think I've seen an issue about this on ocaml/ocaml.   It would reduce the amount of runtime internals you need in ocaml-multicont. /cc @kayceesrk

**Comment by dhil** [comment]
My pleasure.

It would quite possibly make sense. I can open a PR on ocaml/ocaml after discussing with KC.

File: packages/multicont/multicont.1.0.1/opam [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  "This library provides facilities for programming with multi-shot continuations in OCaml"
+maintainer: ["Daniel Hillerström"]
+authors: ["Daniel Hillerström"]
+license: "MIT"
+tags: ["multi-shot continuations" "effect handlers"]
+homepage: "https://github.com/dhil/ocaml-multicont"
+bug-reports: "https://github.com/dhil/ocaml-multicont/issues"
+depends: [
+  "ocaml" { >= "5.0.0"}
+  "dune" {>= "3.8" &  >= "3.8.0"}
```
Comment (avsm) [inline packages/multicont/multicont.1.0.1/opam:13]:
```suggestion
  "dune" {>= "3.8"}
```

**Comment by mseri** [comment]
There is still some work to do:
```

#=== ERROR while compiling multicont.1.0.1 ====================================#
# context              2.2.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.5.0.0 | pinned(https://github.com/dhil/ocaml-multicont/releases/download/v1.0.1/multicont-1.0.1.tbz)
# path                 ~/.opam/5.0/.opam-switch/build/multicont.1.0.1
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p multicont -j 31 @install
# exit-code            1
# env-file             ~/.opam/log/multicont-7-17bd90.env
# output-file          ~/.opam/log/multicont-7-17bd90.out
### output ###
# File "lib/dune", line 1, characters 0-319:
#  1 | (library
#  2 |  (name multicont)
#  3 |  (public_name multicont)
# ....
# 12 |   (mode native)
# 13 |   (flags :standard (:include c_native_flags.sexp))
# 14 |   (names multicont_stubs)))
# (cd _build/default && /home/opam/.opam/5.0/bin/ocamlopt.opt -w -40 -g -shared -linkall -I lib -o lib/multicont.cmxs lib/multicont.cmxa)
# /usr/bin/ld: cannot find -lmulticont_stubs_native: No such file or directory
# collect2: error: ld returned 1 exit status
# File "caml_startup", line 1:
# Error: Error during linking (exit code 1)
```
And when only the bytecode is available
```
#=== ERROR while compiling multicont.1.0.1 ====================================#
# context              2.2.0~alpha~dev | linux/x86_32 | ocaml-base-compiler.5.0.0 | pinned(https://github.com/dhil/ocaml-multicont/releases/download/v1.0.1/multicont-1.0.1.tbz)
# path                 ~/.opam/5.0/.opam-switch/build/multicont.1.0.1
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p multicont -j 31 @install
# exit-code            1
# env-file             ~/.opam/log/multicont-7-440a72.env
# output-file          ~/.opam/log/multicont-7-440a72.out
### output ###
# File "lib/dune", line 4, characters 13-19:
# 4 |  (modes byte native)
#                  ^^^^^^
# Error: this mode isn't available
```

**Comment by dhil** [comment]
> There is still some work to do:
> 
> ```
> 
> #=== ERROR while compiling multicont.1.0.1 ====================================#
> # context              2.2.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.5.0.0 | pinned(https://github.com/dhil/ocaml-multicont/releases/download/v1.0.1/multicont-1.0.1.tbz)
> # path                 ~/.opam/5.0/.opam-switch/build/multicont.1.0.1
> # command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p multicont -j 31 @install
> # exit-code            1
> # env-file             ~/.opam/log/multicont-7-17bd90.env
> # output-file          ~/.opam/log/multicont-7-17bd90.out
> ### output ###
> # File "lib/dune", line 1, characters 0-319:
> #  1 | (library
> #  2 |  (name multicont)
> #  3 |  (public_name multicont)
> # ....
> # 12 |   (mode native)
> # 13 |   (flags :standard (:include c_native_flags.sexp))
> # 14 |   (names multicont_stubs)))
> # (cd _build/default && /home/opam/.opam/5.0/bin/ocamlopt.opt -w -40 -g -shared -linkall -I lib -o lib/multicont.cmxs lib/multicont.cmxa)
> # /usr/bin/ld: cannot find -lmulticont_stubs_native: No such file or directory
> # collect2: error: ld returned 1 exit status
> # File "caml_startup", line 1:
> # Error: Error during linking (exit code 1)
> ```

Interesting. I am unable to reproduce this error on my local machines. Though, my own CI did produce this error too at some point (c.f. https://github.com/dhil/ocaml-multicont/actions/runs/5356156222/jobs/9715536594). Looking at my own CI history it looks like a rather random occurrence, which makes me think this may be due some concurrency bug (e.g. data race).


> And when only the bytecode is available
> 
> ```
> #=== ERROR while compiling multicont.1.0.1 ====================================#
> # context              2.2.0~alpha~dev | linux/x86_32 | ocaml-base-compiler.5.0.0 | pinned(https://github.com/dhil/ocaml-multicont/releases/download/v1.0.1/multicont-1.0.1.tbz)
> # path                 ~/.opam/5.0/.opam-switch/build/multicont.1.0.1
> # command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p multicont -j 31 @install
> # exit-code            1
> # env-file             ~/.opam/log/multicont-7-440a72.env
> # output-file          ~/.opam/log/multicont-7-440a72.out
> ### output ###
> # File "lib/dune", line 4, characters 13-19:
> # 4 |  (modes byte native)
> #                  ^^^^^^
> # Error: this mode isn't available
> ```

Ah bumper. I will have to ask some dune-expert about this.

**Comment by raphael-proust** [comment]
I've restarted the failing checks because you seem to suggest some errors are random.

As for the `(modes byte native)` I'd suggest to do `(modes byte best)` because `best` is going to be `native` if available and `bytes` otherwise. I hope that `dune` remove duplicate modes internally so it doesn't build the byte part twice on non-native platforms. Any `dune` expert to chip in?

**Comment by dhil** [comment]
It looks like the `(modes byte best)` trick works. Also, restricting the number of parallel jobs to `1` seems have resolved the linking error, which strengthens my belief that there may be a data race.

**Comment by mseri** [comment]
Fantastic, thank you all!

**Pull request merged**